### PR TITLE
Improve amp-bind documentation

### LINKED ISSUE #8392: amp-bind: Make docs better(er)
- We should get `amp-bind.md` good enough such that cross-referencing ampbyexample.com isn't necessary for new users.
- Restore bindable properties/attributes matrix in light of #8272.

/cc @ericlindley-g @kmh287

### PR DESCRIPTION
**Pull request by choumx**
Fixes #8392.

- Rework the "Overview" section.
- Update the list of whitelisted functions and add the new perf constraints.
- Add a more complex example based off of the ABE sample.

### CODE REVIEW

**Comment by ericlindley-g** [comment]
Excellent, from my perspective — super thorough and clear.

File: extensions/amp-bind/amp-bind.md [hunk 1]
```diff
@@ -37,36 +37,101 @@ limitations under the License.
...
+Tapping the button causes the `<p>` element's text to change from "Hello World" to "Hello amp-bind".
+
+## How does it work?
+
+`amp-bind` has three main components:
+
+1. [State](#state)
+  - In the example above, the state is empty before tapping the button and `{foo: 'amp-bind'}` after tapping the button.
+2. [Expressions](#expressions)
+  - The example above has a single expression, `'Hello' + foo`, which concatenates the string literal `'Hello'` and the variable state `foo`.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Really small nit: the string literal is `'Hello '`

File: extensions/amp-bind/amp-bind.md [hunk 2]
```diff
@@ -37,36 +37,101 @@ limitations under the License.
...
+3. [Bindings](#bindings)
+  - The example above has a single binding, `[text]`, which causes the `<p>` element to update its text every time the bound expression's value changes.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: text content?

File: extensions/amp-bind/amp-bind.md [hunk 3]
```diff
@@ -37,36 +37,101 @@ limitations under the License.
...
+
+Note that `amp-bind` does no evaluate expressions on page load, so there's no risk of unexpected content jumping. `amp-bind` also takes special care to ensure speed, security and performance on AMP pages.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: nit: "does not"

Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: How about: "so there's no risk of content jumping unexpectedly"?

File: extensions/amp-bind/amp-bind.md [hunk 4]
```diff
@@ -76,48 +141,76 @@ The state can be initialized with the `amp-state` component:
...
 | [Node.textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) | `[text]` | Supported on most text elements.
-| [CSS Classes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class) | `[class]` | Expression result must be a space-delimited string.
-| Element-specific attribute | `[<attr>]` | Only whitelisted attributes are supported.<br>Boolean expression results toggle boolean attributes.
+| [CSS classes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class) | `[class]` | Expression result must be a space-delimited string.
+| Size of [AMP elements](https://www.ampproject.org/docs/reference/components) | `[width]`<br>`[height]` | Changes the width and/or height of the AMP element. |
+| Element-specific attribute | [Various](#element-specific-attributes). | Boolean expression results toggle boolean attributes.
 
-- For security reasons, binding to `innerHTML` is disallowed
-- All attribute bindings are sanitized for unsafe URL protocols (e.g. `javascript:`)
+- For security reasons, binding to `innerHTML` is disallowed.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: We prevent binding to a lot of attributes. Do you think this one is worth explicitly calling out?

File: extensions/amp-bind/amp-bind.md [hunk 5]
```diff
@@ -76,48 +141,76 @@ The state can be initialized with the `amp-state` component:
...
 
-| Type | Syntax | Details |
+| Type | Attribute | Details |
 | --- | --- | --- |
 | [Node.textContent](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent) | `[text]` | Supported on most text elements.
-| [CSS Classes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class) | `[class]` | Expression result must be a space-delimited string.
-| Element-specific attribute | `[<attr>]` | Only whitelisted attributes are supported.<br>Boolean expression results toggle boolean attributes.
+| [CSS classes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/class) | `[class]` | Expression result must be a space-delimited string.
+| Size of [AMP elements](https://www.ampproject.org/docs/reference/components) | `[width]`<br>`[height]` | Changes the width and/or height of the AMP element. |
+| Element-specific attribute | [Various](#element-specific-attributes). | Boolean expression results toggle boolean attributes.
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Maybe remove this row in the table and leave introducing component-specific attributes for the next section?

File: extensions/amp-bind/amp-bind.md [hunk 6]
```diff
@@ -76,48 +141,76 @@ The state can be initialized with the `amp-state` component:
...
-There are also special bindable attributes without non-bindable counterparts:
+(*) Denotes bindable attributes that don't have a non-bindable counterpart.
 
-| Component | Attribute | Details | Example |
-| --- | --- | --- | --- |
-| amp-carousel[type=slides] | `[slide]` | The currently displayed slide index. | [Linked carousels](https://ampbyexample.com/advanced/image_galleries_with_amp-carousel/#linking-carousels-with-amp-bind)
-| amp-selector | `[selected]` | The `option` attribute values of the currently selected children elements. | [Linked carousels](https://ampbyexample.com/advanced/image_galleries_with_amp-carousel/#linking-carousels-with-amp-bind)
+| Component | Attribute(s) | Behavior |
+| --- | --- | --- |
+| `<amp-brightcove>` | `[data-account]`<br>`[data-embed]`<br>`[data-player]`<br>`[data-player-id]`<br>`[data-playlist-id]`<br>`[data-video-id]` | Changes the displayed Brightcove video. |
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Is this behavior accurate to changing any of these attributes? Maybe it's worth just linking to the amp-brightcove page on ampproject like you do elsewhere?

File: extensions/amp-bind/amp-bind.md [hunk 7]
```diff
@@ -76,48 +141,76 @@ The state can be initialized with the `amp-state` component:
...
+| `<amp-carousel type=slides>` | `[slide]`* | Changes the currently displayed slide index. [See an example](https://ampbyexample.com/advanced/image_galleries_with_amp-carousel/#linking-carousels-with-amp-bind).
+| `<amp-iframe>` | `[src]` | Changes the iframe's source URL. |
+| `<amp-img>` | `[alt]`<br>`[attribution]`<br>`[src]`<br>`[srcset]` | Various. See [amp-img attributes](https://www.ampproject.org/docs/reference/components/media/amp-img#attributes). |
+| `<amp-selector>` | `[selected]`* | Changes the currently selected children element(s)<br>identified by their `option` attribute values. Supports a comma-separated list of values for multiple selection. [See an example](https://ampbyexample.com/advanced/image_galleries_with_amp-carousel/#linking-carousels-with-amp-bind).
+| `<amp-video>` | `[alt]`<br>`[attribution]`<br>`[controls]`<br>`[loop]`<br>`[poster]`<br>`[preload]`<br>`[src]` | Various. See [amp-video attributes](https://www.ampproject.org/docs/reference/components/media/amp-video#attributes). |
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Suggestion: here and elsewhere, how would you feel about replacing 


"Various. See ${ELEMENT} attributes"

with 

"See corresponding ${ELEMENT} attributes"

File: extensions/amp-bind/amp-bind.md [hunk 8]
```diff
@@ -76,48 +141,76 @@ The state can be initialized with the `amp-state` component:
...
+- Expressions may only access the containing document's [state](#state).
+- Expressions **do not** have access to globals like `window` or `document`.
+- Only [whitelisted functions](#whitelisted-functions) are allowed.
+- Custom functions, classes and some control flow statements (e.g. `for`) are disallowed.
+- Undefined variables and array-index-out-of-bounds return `null` instead of `undefined` or throwing errors.
+- A single expression is currently capped at 50 operands for performance reasons. Please [contact us](https://github.com/ampproject/amphtml/issues/new) if this is insufficient for your use case.
+
+#### Whitelisted functions
+
+| Object type | Function(s) | Example |
```
Comment (kmh287) [inline extensions/amp-bind/amp-bind.md]: Also add the whitelisted math functions:

abs, ceil, floor, max, min, random, round, sign

File: extensions/amp-bind/amp-bind.md (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

File: extensions/amp-bind/amp-bind.md (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: I think "text" is better for the simple overview. We mention `Node.textContent` in the details section.

File: extensions/amp-bind/amp-bind.md (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

File: extensions/amp-bind/amp-bind.md (hunk 4 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: I imagine it'd be a common question. Other frameworks have similar callouts for `innerHTML`.

File: extensions/amp-bind/amp-bind.md (hunk 6 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Yep, all of those params are video identifiers.

File: extensions/amp-bind/amp-bind.md (hunk 7 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

File: extensions/amp-bind/amp-bind.md (hunk 8 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Done.

**Comment by choumx** [comment]
> Do you think it's worth adding a section on how to debug amp-bind?

Good idea. Probably also useful to have a section describing different types of errors and a FAQs section. Filed #8549.

File: extensions/amp-bind/amp-bind.md (hunk 5 above)
Comment (choumx) [inline extensions/amp-bind/amp-bind.md]: Oops, missed this comment. This format makes sense to me though you may be right. I'll consider this in the next PR.

**Comment by choumx** [comment]
Merging since Barb is out this week.

**Pull request merged**